Add front end support for type matching

AustinMroz opened the pull request:
This PR implements front end logic to handle MatchType inputs and outputs.
See  comfyanonymous/ComfyUI#10644

This allows for the implementation of nodes such as a "switch node"  where input types change based on the connections made.
![switch-node](https://github.com/user-attachments/assets/090515ba-484c-4295-b7b3-204b0c72fc4a)

As part of this implementation, significant cleanup is being performed in the reroute code. Extra testing will be required to make sure these changes don't introduce regressions.

┆Issue is synchronized with this [Notion page](https://www.notion.so/PR-6582-Add-front-end-support-for-type-matching-2a16d73d36508189b042cd23f82a332e) by [Unito](https://www.unito.io)

AustinMroz committed:
Implement a new switch node

AustinMroz committed:
WIP link cleanup

AustinMroz committed:
Type cleanup

AustinMroz committed:
Trigger onConnectionChange on connected outputs

AustinMroz committed:
Clean up typechecks, fix output typing

AustinMroz committed:
Set dirty canvas on outputs changed

AustinMroz committed:
Fix ts-expect-error on intersection

AustinMroz committed:
Reroute cleanup

AustinMroz committed:
Further reroute cleanup

AustinMroz committed:
More minor syntax tweaks

AustinMroz committed:
Adopt V3 schema implementation

AustinMroz committed:
Isolate timeout handling

AustinMroz committed:
fix template typing

AustinMroz committed:
move file from switchNode to matchGroup

AustinMroz committed:
Fix subgraphOutput typing

AustinMroz committed:
Fix missed index update

AustinMroz committed:
Properly use matchtype groups for outputs

AustinMroz committed:
Fix logic for >2 inputs in a group

AustinMroz committed:
Fix link display color

I think updating this to be the color of the combiend output type is
most correct, but implementing this seems logically excessive and I
don't think it will ever be relevant

AustinMroz committed:
Improve SubgraphOutput.disconnect

Now cleans up input link reference and calls onConnectionsChange

AustinMroz committed:
Fix wrong slotType and index being used

AustinMroz committed:
Fix reroute bugs

clone was causing infinite recursion, missing inputs resulted in broken
links

AustinMroz committed:
Fix diverged logic in reroute code.

christian-byrne commented on src/extensions/core/matchType.ts:
```diff
@@ -0,0 +1,151 @@
+  )
+  if (combinedSet.size == 0) return undefined
+  return [...combinedSet].join(',')
+}
```
nit: I wonder if we can make this code more expressive so people can read it quickly and understand everything that it's doing.

Example:

```typescript
import { intersection, without, split } from 'es-toolkit';

/** comment explain the semantics of what `undefined` return value represents, etc. **/
export function combineTypes(...types: ISlotType[]): ISlotType | undefined {
  if (!isStrings(types)) return undefined

  const withoutWildcards = without(types, '*')

  const allWereWildcards = withoutWildcards.length === 0
  if (allWereWildcards) return '*'

  const allTypes = withoutWildcards.map(t => split(t, ','))
  const commonTypes = intersection(...allTypes) // order follows the first array

  const hasCommonTypes = !!commonTypes.length
  return hasCommonTypes ? commonTypes.join(',') : undefined
}
```

AustinMroz replied:
Agreed.

Was real excited to see es-toolkit intersection, but it seems to have the same 'limit 2 inputs' which requires the ugly `reduce`.

I've opted to write my own intersection implementation which does handle variable number of inputs. I think it covers the same need of segmenting out the complex logic so it's clear how types are combined.

AustinMroz committed:
Refactor type combination logic

AustinMroz committed:
Remove debounce for type changes

Connecting a new link over an existing link triggers a disconnect event
before performing a new connection event. This creates intermediate
states which could potentially be invalid.

However, disconnecting inputs can only "loosen" the connection logic for
other inputs and outputs. This means that the temporary disconnection
when swapping links can never create an invalid state.

As a result, this debounce is neither required nor useful.

christian-byrne reviewed (approved):
I think we can merge this.

For future, the things I would consider:

- Explaining choice between changing litegraph directly vs. using monkey-patches
- Running the code through an LLM and asking how it could be made more expressive/maintainable (e.g., "how can I refactor or document this code such that a new hire could read it and fix a bug related to it?")
- Trying to see if there's any ways to make minor adjustments that would make the code better match the style of codebase a bit more (minor)

From a technical perspective, I think it is great code and I definitely want to see the feature get in the users' hands as soon as possible.
